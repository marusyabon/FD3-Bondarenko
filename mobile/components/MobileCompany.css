.MobileCompany
{
  border: double gray 3px;
  width: 330px;
  padding: 3px 7px;
  background-color: lightyellow;
}

.MobileCompanyName
{
  font-weight: bold;
}

.MobileCompanyClients
{
  padding-top: 5px;
}

.Buttons {
  padding: 10px 0;
}

.Buttons input {
  width: 105px;
  padding: 3px 0;
  text-align: center;
}

.Buttons input:not(:first-child) {
  margin-left: 5px;
}

.EditClient {
  margin: 20px 0;
}

.EditClient input {
  margin-top: 5px;
  margin-right: 5px;
}