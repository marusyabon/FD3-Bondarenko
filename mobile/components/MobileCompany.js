import React from 'react';
import PropTypes from 'prop-types';

import MobileClient from './MobileClient';

import './MobileCompany.css';

class MobileCompany extends React.PureComponent {

  static propTypes = {
    clients:PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        name: PropTypes.string.isRequired,
        secondName: PropTypes.string.isRequired,
        balance: PropTypes.number.isRequired,
      })
    )
  };

  state = {
    clients: this.props.clients,
    filteredClients: this.props.clients,
    currentClientId: null,
    workMode: 0
  };

  defineNewId = () => {
    let maxId = this.state.clients.reduce((isMax, item) => (isMax > item ? isMax : item) );
    return maxId.id+1
  }

  currentClient = null

  addClient = () => {
    this.state.workMode !== 1 && 
      this.setState({
        workMode: 1
      });
    this.currentClient = {id:this.defineNewId(), name:"", secondName:"", balance:0}
  }

  changeValue = (EO) => { 

    let key = EO.target.name;
    let val = EO.target.value;

    key == "balance"
    ? this.currentClient[key] = +val
    : this.currentClient[key] = val;
  }

  cancelEdit = () => {
    this.setState({
      workMode: 0
    });
  }

  editClient = (id) => {
    let clientEdited = this.state.clients.filter(c => {
      return c.id == id
    });
    this.currentClient = clientEdited[0];

    this.setState({
      workMode: 2,
      currentClientId: id
    });

    console.log(this.state.currentClientId)

  }

  saveValue = (EO) => {
    EO.preventDefault();

    if(this.state.workMode == 1) {
      let newFilteredClients = [...this.state.clients, this.currentClient];
      
      this.setState({
        filteredClients: [...newFilteredClients],
        clients: [...newFilteredClients],
        workMode: 0
      });
    }

    if(this.state.workMode == 2) {
      let newFilteredClients = [...this.state.clients];
      newFilteredClients.forEach((c, i) => {
        c.id == this.currentClient.id && (newFilteredClients[i] = this.currentClient)
      });

      this.setState({
        filteredClients: newFilteredClients,
        clients: newFilteredClients,
        workMode: 0
      });
    }

    console.log(this.state.workMode)
  }
  
  deleteClient = (id) => {
    let newFilteredClients = [...this.state.clients];
    newFilteredClients.forEach((c, i) => {
      if (c.id == id) {
        newFilteredClients.splice(i, 1);
        return;
      }
    });
    this.setState({
      filteredClients: newFilteredClients,
      clients: newFilteredClients
    })
  }

  showActive = () => {
    if (this.state.filter != "showActive") {
      let newFilteredClients = this.state.clients.filter(c => {
        return c.balance > 0;
      });
      this.setState({
        filteredClients: newFilteredClients,
        filter: "showActive"
      })
    }
  }

  showBlocked = () => {
    if (this.state.filter != "showBlocked") {
      let newFilteredClients = this.state.clients.filter(c => {
        return c.balance <= 0;
      });
      this.setState({
        filteredClients: newFilteredClients,
        filter: "showBlocked"
      })
    }
  }

  showAll = () => {
    if (this.state.filter != "showAll") {
      this.setState({
        filteredClients: [...this.state.clients],
        filter: "showAll"
      })
    }
  }
  
  render() {

    console.log("MobileCompany render");

    var clientsCode=this.state.filteredClients.map( client => {

      return (
        <MobileClient 
          key={client.id} 
          clientInfo={client} 
          cbDeleted={this.deleteClient}
          cbEdited={this.editClient}
        />);
      }
    );

    return (
      <div className='MobileCompany'>
        <div className="Buttons">
          <input type="button" value="Show Active" onClick={this.showActive}/>
          <input type="button" value="Show Blocked" onClick={this.showBlocked}/>
          <input type="button" value="Clear filter" onClick={this.showAll}/>
        </div>
        <div className='MobileCompanyName'>Mobile company</div>
        <div className='MobileCompanyClients'>
          {clientsCode}
        </div>

        {
          this.state.workMode == 0 && 
          (
            <div className="Buttons">
              <input type="button" value="Add" onClick={this.addClient}/>
            </div>
          )
        }

        {
          this.state.workMode == 1 && 
          (
          <form className="EditClient">
            <input type="text" name="name" placeholder="Name" onChange={this.changeValue}/>
            <input type="text" name="secondName" placeholder="Second Name" onChange={this.changeValue}/>
            <input type="text" name="balance" placeholder="Balance" onChange={this.changeValue}/>

            <div>
              <input type="submit" value="Save" onClick={this.saveValue}/>
              <input type="button" value="Cancel" onClick={this.cancelEdit}/>
            </div>
            
          </form>
          )

        }

        {
          this.state.workMode == 2 && 
          (          
          <form className="EditClient">
            <input type="text" name="name" defaultValue={this.currentClient.name} onChange={this.changeValue}/>
            <input type="text" name="secondName" defaultValue={this.currentClient.secondName} onChange={this.changeValue}/>
            <input type="text" name="balance" defaultValue={this.currentClient.balance} onChange={this.changeValue}/>

            <div>
              <input type="submit" value="Save" onClick={this.saveValue}/>
              <input type="button" value="Cancel" onClick={this.cancelEdit}/>
            </div>
            
          </form>
          )

        }

      </div>
    );

  }

}

export default MobileCompany;
