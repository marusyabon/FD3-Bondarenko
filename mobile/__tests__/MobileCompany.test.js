"use strict";

import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, configure  } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';


import MobileCompany from '../components/MobileCompany';

configure({ adapter: new Adapter() });

describe('>>>>MOBILE COMPANY functions', () => {

  let clientsArr=[ 
    {id:101, secondName:"Иванов", name:"Иван", balance:200}, 
    {id:105, secondName:"Сидоров", name:"Сидор", balance:250}, 
    {id:110, secondName:"Петров", name:"Пётр", balance:180},
    {id:120, secondName:"Григорьев", name:"Григорий", balance:220},
    {id:121, secondName:"Павлов", name:"Павел", balance:-20},
    {id:122, secondName:"Сергеев", name:"Сергей", balance:-320},
    {id:123, secondName:"Юрьев", name:"Юрий", balance:0},
  ];  
  
  let wrapper;
  let company = <MobileCompany clients={clientsArr}/>;

  beforeEach(()=>{
    wrapper = shallow(company)
  })

  it('ADD CLIENT', () => {
    wrapper.instance().addClient();
    wrapper.instance().saveValue();

    let clients = wrapper.state('clients');

    expect(clients.length).toEqual(8);
    expect(clients[clients.length-1].id).toEqual(124);
  });

  it('CHANGE CLIENT', () => {
    wrapper.instance().editClient(120);
    wrapper.instance().currentClient.secondName = 'Власов';
    wrapper.instance().currentClient.balance = 150;

    let editingClient = wrapper.state('clients')[3];

    expect(editingClient.name).toEqual('Григорий');
    expect(editingClient.secondName).toEqual('Власов');
    expect(editingClient.balance).toEqual(150);
  });

  it('DELETE CLIENT', () => {
    wrapper.instance().deleteClient(101);
    expect(wrapper.state('clients').length).toEqual(6);
    expect(wrapper.state('clients')[0].id).toEqual(105);
  });

  it('SHOW ACTIVE', () => {
    wrapper.instance().showActive();
    expect(wrapper.state('filteredClients').length).toEqual(4);
    expect(wrapper.state('filteredClients')[2].balance).toEqual(180);
  });

  it('SHOW BLOCKED', () => {
    wrapper.instance().showBlocked();
    expect(wrapper.state('filteredClients').length).toEqual(3);
    expect(wrapper.state('filteredClients')[0].name).toEqual('Павел');
  });

  //SNAPSHOT
  let component = renderer.create(company);

  it('CHECK RENDER', () => {
    let componentTree = component.toJSON();
    expect(componentTree).toMatchSnapshot();
  });
})