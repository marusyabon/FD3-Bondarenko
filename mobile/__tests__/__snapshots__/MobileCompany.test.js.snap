// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>>>MOBILE COMPANY functions CHECK RENDER 1`] = `
<div
  className="MobileCompany"
>
  <div
    className="Buttons"
  >
    <input
      onClick={[Function]}
      type="button"
      value="Show Active"
    />
    <input
      onClick={[Function]}
      type="button"
      value="Show Blocked"
    />
    <input
      onClick={[Function]}
      type="button"
      value="Clear filter"
    />
  </div>
  <div
    className="MobileCompanyName"
  >
    Mobile company
  </div>
  <div
    className="MobileCompanyClients"
  >
    <div
      className="MobileClient"
      id={101}
    >
      <span
        className="MobileClientName"
      >
        Иван Иванов
      </span>
      <span
        className="MobileClientBalance"
      >
        200
      </span>
      <span
        className="MobileClientBtn"
      >
        <input
          onClick={[Function]}
          type="button"
          value="Edit"
        />
      </span>
      <span
        className="MobileClientBtn"
      >
        <input
          onClick={[Function]}
          type="button"
          value="Delete"
        />
      </span>
    </div>
    <div
      className="MobileClient"
      id={105}
    >
      <span
        className="MobileClientName"
      >
        Сидор Сидоров
      </span>
      <span
        className="MobileClientBalance"
      >
        250
      </span>
      <span
        className="MobileClientBtn"
      >
        <input
          onClick={[Function]}
          type="button"
          value="Edit"
        />
      </span>
      <span
        className="MobileClientBtn"
      >
        <input
          onClick={[Function]}
          type="button"
          value="Delete"
        />
      </span>
    </div>
    <div
      className="MobileClient"
      id={110}
    >
      <span
        className="MobileClientName"
      >
        Пётр Петров
      </span>
      <span
        className="MobileClientBalance"
      >
        180
      </span>
      <span
        className="MobileClientBtn"
      >
        <input
          onClick={[Function]}
          type="button"
          value="Edit"
        />
      </span>
      <span
        className="MobileClientBtn"
      >
        <input
          onClick={[Function]}
          type="button"
          value="Delete"
        />
      </span>
    </div>
    <div
      className="MobileClient"
      id={120}
    >
      <span
        className="MobileClientName"
      >
        Григорий Григорьев
      </span>
      <span
        className="MobileClientBalance"
      >
        220
      </span>
      <span
        className="MobileClientBtn"
      >
        <input
          onClick={[Function]}
          type="button"
          value="Edit"
        />
      </span>
      <span
        className="MobileClientBtn"
      >
        <input
          onClick={[Function]}
          type="button"
          value="Delete"
        />
      </span>
    </div>
    <div
      className="MobileClient"
      id={121}
    >
      <span
        className="MobileClientName"
      >
        Павел Павлов
      </span>
      <span
        className="MobileClientBalance"
      >
        -20
      </span>
      <span
        className="MobileClientBtn"
      >
        <input
          onClick={[Function]}
          type="button"
          value="Edit"
        />
      </span>
      <span
        className="MobileClientBtn"
      >
        <input
          onClick={[Function]}
          type="button"
          value="Delete"
        />
      </span>
    </div>
    <div
      className="MobileClient"
      id={122}
    >
      <span
        className="MobileClientName"
      >
        Сергей Сергеев
      </span>
      <span
        className="MobileClientBalance"
      >
        -320
      </span>
      <span
        className="MobileClientBtn"
      >
        <input
          onClick={[Function]}
          type="button"
          value="Edit"
        />
      </span>
      <span
        className="MobileClientBtn"
      >
        <input
          onClick={[Function]}
          type="button"
          value="Delete"
        />
      </span>
    </div>
    <div
      className="MobileClient"
      id={123}
    >
      <span
        className="MobileClientName"
      >
        Юрий Юрьев
      </span>
      <span
        className="MobileClientBalance"
      >
        0
      </span>
      <span
        className="MobileClientBtn"
      >
        <input
          onClick={[Function]}
          type="button"
          value="Edit"
        />
      </span>
      <span
        className="MobileClientBtn"
      >
        <input
          onClick={[Function]}
          type="button"
          value="Delete"
        />
      </span>
    </div>
  </div>
  <div
    className="Buttons"
  >
    <input
      onClick={[Function]}
      type="button"
      value="Add"
    />
  </div>
</div>
`;
